File: mobile/lib/features/jobs/models/job.dart
class IndexJob {
  const IndexJob({
    required this.id,
    required this.jobType,
    required this.status,
    required this.progress,
    this.documentId,
    this.retryOfJobId,
    this.message,
    this.errorMessage,
    this.createdAt,
    this.updatedAt,
  });

  final String id;
  final String jobType;
  final String status;
  final int progress;
  final String? documentId;
  final String? retryOfJobId;
  final String? message;
  final String? errorMessage;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  bool get isFinished =>
      status == 'success' || status == 'failed' || status == 'cancelled';

  bool get canCancel => status == 'pending' || status == 'running';

  bool get canRetry => status == 'failed' || status == 'cancelled';

  factory IndexJob.fromJson(Map<String, dynamic> json) {
    return IndexJob(
      id: json['id'] as String,
      jobType: json['job_type'] as String,
      status: json['status'] as String,
      progress: json['progress'] as int? ?? 0,
      documentId: json['document_id'] as String?,
      retryOfJobId: json['retry_of_job_id'] as String?,
      message: json['message'] as String?,
      errorMessage: json['error_message'] as String?,
      createdAt: json['created_at'] == null
          ? null
          : DateTime.parse(json['created_at'] as String),
      updatedAt: json['updated_at'] == null
          ? null
          : DateTime.parse(json['updated_at'] as String),
    );
  }
}

File: backend/app/models/job.py
from datetime import datetime
from enum import Enum
from uuid import UUID as PyUUID
from uuid import uuid4

from sqlalchemy import DateTime, ForeignKey, Index, Integer, String, Text, func, text
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base


class JobType(str, Enum):
    INDEX_DOCUMENT = "index_document"
    REBUILD_EMBEDDINGS = "rebuild_embeddings"
    REBUILD_ALL_EMBEDDINGS = "rebuild_all_embeddings"


class JobStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    CANCEL_REQUESTED = "cancel_requested"
    CANCELLED = "cancelled"
    SUCCESS = "success"
    FAILED = "failed"


class Job(Base):
    __tablename__ = "jobs"
    __table_args__ = (
        Index("ix_jobs_user_status", "user_id", "status"),
        Index("ix_jobs_updated_at_id", "updated_at", "id"),
        Index(
            "ix_jobs_recoverable_created_at_id",
            "created_at",
            "id",
            postgresql_where=text(
                "status IN ('pending', 'cancel_requested', 'running')"
            ),
        ),
        Index(
            "ix_jobs_user_updated_created_id",
            "user_id",
            "updated_at",
            "created_at",
            "id",
        ),
    )

    id: Mapped[PyUUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    user_id: Mapped[PyUUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="RESTRICT"),
        index=True,
        nullable=False,
    )
    document_id: Mapped[PyUUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("documents.id", ondelete="SET NULL"),
        index=True,
    )
    retry_of_job_id: Mapped[PyUUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("jobs.id", ondelete="SET NULL"),
        index=True,
    )
    run_token: Mapped[PyUUID | None] = mapped_column(UUID(as_uuid=True))
    job_type: Mapped[str] = mapped_column(String(64), nullable=False)
    configuration_fingerprint: Mapped[str | None] = mapped_column(
        String(192),
        index=True,
    )
    status: Mapped[str] = mapped_column(String(32), nullable=False, default=JobStatus.PENDING.value)
    progress: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    message: Mapped[str | None] = mapped_column(String(255))
    error_message: Mapped[str | None] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )
